File: scripts/camera/test-jpeg-rotation.cc
/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * deliverJpeg 旋转逻辑的【真编译 + 真运行】自测（无第三方依赖）。
 *
 * 为什么要有这个：scripts/camera/test-jpeg-rotation.py 做的是【源码级】断言
 * —— 它保证 Session.cpp 里的表达式就是我们想的那几行；本文件做的是【运行级】
 * 检查 —— 把同一段表达式编译起来跑，用按 libyuv 契约实现的替身，验证调用
 * 形状真的对：实参顺序、源/目标 stride、90/270 的宽高互换、像素真的搬对了。
 *
 * 出 bug 的地方几乎都在调用形状上（比如目标 stride 顺手写成 dstW*4，画面会被
 * 斜切成条），而那种错在纯数学的模型里看不出来 —— 必须让数据真的流一遍。
 *
 * ⚠️ 这里用到的 libyuv 是【替身】，语义严格照 include/libyuv/rotate.h 的注释
 *    写死为顺时针：
 *       kRotate90  = 90,   // Rotate 90 degrees clockwise.
 *       kRotate270 = 270,  // Rotate 270 degrees clockwise.
 *    真机上换成真库即可，调用点不变。
 *
 * 构建与运行（WSL / 任意 Linux）：
 *     g++ -std=c++20 -Wall -Wextra -Werror -O1 \
 *         -o /tmp/t scripts/camera/test-jpeg-rotation.cc && /tmp/t
 */

#include <cstdint>
#include <cstdio>
#include <cstring>
#include <vector>

/* ───────────── libyuv 替身（只实现我们用到的三个函数） ───────────── */

namespace libyuv {

enum RotationMode {
	kRotate0 = 0,     /* No rotation. */
	kRotate90 = 90,   /* Rotate 90 degrees clockwise. */
	kRotate180 = 180, /* Rotate 180 degrees. */
	kRotate270 = 270, /* Rotate 270 degrees clockwise. */
};

enum FilterMode {
	kFilterNone = 0,
	kFilterBilinear = 2,
};

/* "ARGB" 在内存里其实是 B,G,R,A —— 与 libcamera 的 RGB888(=BGR24) 和
 * libjpeg 的 JCS_EXT_BGRA 都对得上，替身照这个布局来。 */
inline int RGB24ToARGB(const uint8_t *s, int ss, uint8_t *d, int ds, int w, int h)
{
	if (!s || !d || w <= 0 || h <= 0)
		return -1;
	for (int y = 0; y < h; y++) {
		for (int x = 0; x < w; x++) {
			const uint8_t *sp = s + static_cast<size_t>(y) * ss + x * 3;
			uint8_t *dp = d + static_cast<size_t>(y) * ds + x * 4;
			dp[0] = sp[0];
			dp[1] = sp[1];
			dp[2] = sp[2];
			dp[3] = 0xFF;
		}
	}
	return 0;
}

/* 最近邻缩放：只为跑通"需要缩放"那条分支，精度不重要。 */
inline int ARGBScale(const uint8_t *s, int ss, int sw, int sh,
		     uint8_t *d, int ds, int dw, int dh, int)
{
	if (!s || !d || sw <= 0 || sh <= 0 || dw <= 0 || dh <= 0)
		return -1;
	for (int y = 0; y < dh; y++) {
		const int sy = y * sh / dh;
		for (int x = 0; x < dw; x++) {
			const int sx = x * sw / dw;
			memcpy(d + static_cast<size_t>(y) * ds + x * 4,
			       s + static_cast<size_t>(sy) * ss + sx * 4, 4);
		}
	}
	return 0;
}

/*
 * 顺时针旋转。width/height 是【源】尺寸 —— 与真库一致：目标 stride 由调用方
 * 按旋转后的宽度给，给错了就会错位，这正是本测试要抓的东西。
 *
 * 用"源像素正向散射"写，避免逆映射那套下标把自己绕进去：
 *   源 (sx,sy) 顺时针 90 度后落在 (dstH-1-sy, sx)      —— 源的上边变成 dst 的右边
 *   180 度落在 (dstW-1-sx, dstH-1-sy)
 *   270 度落在 (sy, dstW-1-sx)
 */
inline int ARGBRotate(const uint8_t *s, int ss, uint8_t *d, int ds,
		      int w, int h, RotationMode mode)
{
	if (!s || !d || w <= 0 || h <= 0)
		return -1;
	const bool swap = mode == kRotate90 || mode == kRotate270;
	const int ow = swap ? h : w;
	const int oh = swap ? w : h;
	for (int sy = 0; sy < h; sy++) {
		for (int sx = 0; sx < w; sx++) {
			int dx = sx, dy = sy;
			switch (mode) {
			case kRotate90:  dx = h - 1 - sy; dy = sx;         break;
			case kRotate180: dx = w - 1 - sx; dy = h - 1 - sy; break;
			case kRotate270: dx = sy;         dy = w - 1 - sx; break;
			default:                                           break;
			}
			memcpy(d + static_cast<size_t>(dy) * ds + dx * 4,
			       s + static_cast<size_t>(sy) * ss + sx * 4, 4);
		}
	}
	(void)ow;
	(void)oh;
	return 0;
}

} /* namespace libyuv */

#define ALOGW(...)                                                             \
	do {                                                                   \
		fprintf(stderr, "    W: " __VA_ARGS__);                        \
		fprintf(stderr, "\n");                                         \
	} while (0)

/* ──────────────── 被测代码：照抄 Session.cpp 的表达式 ──────────────── */

/*
 * ★ 逐字照抄 Session.cpp 的 toLibyuvRotation()。
 *   Python 那边（A2）会断言源码里就是这张表，所以这里和源码任何一边改错
 *   都会有人报错。
 */
static libyuv::RotationMode toLibyuvRotation(int32_t deg)
{
	switch (deg) {
	case 90:
		return libyuv::kRotate90;
	case 180:
		return libyuv::kRotate180;
	case 270:
		return libyuv::kRotate270;
	case 0:
		return libyuv::kRotate0;
	default:
		ALOGW("JPEG_ORIENTATION=%d 不是 0/90/180/270，按不旋转处理", deg);
		return libyuv::kRotate0;
	}
}

/*
 * 照抄 Session::deliverJpeg() 里"缩放 + 旋转"那一段（编码器换成"把最终像素
 * 与尺寸交回来"），表达式一字不改。
 */
struct JpegOut {
	std::vector<uint8_t> buf;   /* 按 stride 紧凑排布 */
	int32_t w = 0, h = 0;
	int channels = 0;           /* 3 = BGR（快捷路径），4 = BGRA */
};

static bool preparePixels(const uint8_t *rgb, int32_t srcWidth_, int32_t srcHeight_,
			  int32_t dstW, int32_t dstH, int32_t jpegOrientation,
			  JpegOut *out)
{
	const libyuv::RotationMode rotMode = toLibyuvRotation(jpegOrientation);
	const bool swapDims = rotMode == libyuv::kRotate90 || rotMode == libyuv::kRotate270;
	const int32_t outW = swapDims ? dstH : dstW;
	const int32_t outH = swapDims ? dstW : dstH;

	const bool needScale = dstW != srcWidth_ || dstH != srcHeight_;
	const bool needRotate = rotMode != libyuv::kRotate0;

	std::vector<uint8_t> argb;      /* dstW×dstH，已缩放、未旋转（BGRA） */
	std::vector<uint8_t> rotated;   /* outW×outH，已旋转（BGRA） */
	const uint8_t *src = rgb;
	int srcStride = srcWidth_ * 3;
	int components = 3;

	if (needScale || needRotate) {
		argb.resize(static_cast<size_t>(dstW) * dstH * 4);

		if (needScale) {
			std::vector<uint8_t> full(static_cast<size_t>(srcWidth_) *
						  srcHeight_ * 4);
			if (libyuv::RGB24ToARGB(rgb, srcWidth_ * 3, full.data(),
						srcWidth_ * 4, srcWidth_, srcHeight_) != 0)
				return false;
			if (libyuv::ARGBScale(full.data(), srcWidth_ * 4, srcWidth_,
					      srcHeight_, argb.data(), dstW * 4, dstW, dstH,
					      libyuv::kFilterBilinear) != 0)
				return false;
		} else {
			if (libyuv::RGB24ToARGB(rgb, srcWidth_ * 3, argb.data(), dstW * 4,
						dstW, dstH) != 0)
				return false;
		}

		if (needRotate) {
			rotated.resize(static_cast<size_t>(outW) * outH * 4);
			if (libyuv::ARGBRotate(argb.data(), dstW * 4, rotated.data(),
					       outW * 4, dstW, dstH, rotMode) != 0)
				return false;
			src = rotated.data();
			srcStride = outW * 4;
		} else {
			src = argb.data();
			srcStride = dstW * 4;
		}
		components = 4;
	}

	if (components == 3) {
		/* 快捷路径：直接拿 RGB24 编码，没有中间拷贝 */
		out->w = dstW;
		out->h = dstH;
		out->channels = 3;
		out->buf.assign(src, src + static_cast<size_t>(srcStride) * dstH);
		return true;
	}

	out->w = outW;
	out->h = outH;
	out->channels = 4;
	out->buf.assign(src, src + static_cast<size_t>(srcStride) * outH);
	return true;
}

/* ─────────────────────────── 断言框架 ─────────────────────────── */

static int g_checks = 0;
static int g_fail = 0;

static void check(bool ok, const char *label)
{
	g_checks++;
	printf("  [%s] %s\n", ok ? "PASS" : "FAIL", label);
	if (!ok)
		g_fail++;
}

/* 源图：像素 (x,y) 的 (B,G,R) = (x+1, y+1, 0x3C) —— 坐标可反查 */
static uint8_t bOf(int x, int y)
{
	(void)y;   /* B 通道只编码 x，与像素布局对称，保留双参是为了调用点好看 */
	return static_cast<uint8_t>((x + 1) & 0xFF);
}
static uint8_t gOf(int x, int y)
{
	(void)x;   /* G 通道只编码 y */
	return static_cast<uint8_t>((y + 1) & 0xFF);
}

static std::vector<uint8_t> makeSource(int w, int h)
{
	std::vector<uint8_t> buf(static_cast<size_t>(w) * h * 3);
	for (int y = 0; y < h; y++)
		for (int x = 0; x < w; x++) {
			uint8_t *p = buf.data() + (static_cast<size_t>(y) * w + x) * 3;
			p[0] = bOf(x, y);
			p[1] = gOf(x, y);
			p[2] = 0x3C;
		}
	return buf;
}

static void testPixels(int srcW, int srcH, int dstW, int dstH, int deg,
		       bool expectScale)
{
	char label[256];
	snprintf(label, sizeof(label),
		 "源 %dx%d → 流 %dx%d、请求旋转 %3d 度：尺寸与每个像素都对",
		 srcW, srcH, dstW, dstH, deg);
	std::vector<uint8_t> src = makeSource(srcW, srcH);
	JpegOut out;
	if (!preparePixels(src.data(), srcW, srcH, dstW, dstH, deg, &out)) {
		check(false, label);
		return;
	}

	const bool swap = (deg == 90 || deg == 270);
	const int expW = swap ? dstH : dstW;
	const int expH = swap ? dstW : dstH;
	const int stride = out.channels == 4 ? 4 : 3;

	/* 期望图：先把源散射成"未旋转的 dst 图"（缩放用最近邻），
	 * 再按顺时针的【正向】映射散射到旋转后的帧里。正向写法不容易绕晕。 */
	const int kW = expW * expH;
	std::vector<int> expB(kW, -1), expG(kW, -1);
	bool ok = (out.w == expW && out.h == expH);

	for (int dy = 0; ok && dy < dstH; dy++) {
		for (int dx = 0; ok && dx < dstW; dx++) {
			const int ux = expectScale ? dx * srcW / dstW : dx;
			const int uy = expectScale ? dy * srcH / dstH : dy;

			int ox = dx, oy = dy;
			switch (deg) {
			case 90:  ox = dstH - 1 - dy; oy = dx;         break;
			case 180: ox = dstW - 1 - dx; oy = dstH - 1 - dy; break;
			case 270: ox = dy;            oy = dstW - 1 - dx; break;
			default:                                        break;
			}
			if (ox < 0 || ox >= expW || oy < 0 || oy >= expH) {
				ok = false;
				break;
			}
			expB[oy * expW + ox] = bOf(ux, uy);
			expG[oy * expW + ox] = gOf(ux, uy);
		}
	}

	for (int i = 0; ok && i < kW; i++) {
		if (expB[i] < 0) {
			ok = false;   /* 有像素没被覆盖到 */
			break;
		}
		const uint8_t *p = out.buf.data() + static_cast<size_t>(i) * stride;
		if (p[0] != static_cast<uint8_t>(expB[i]) ||
		    p[1] != static_cast<uint8_t>(expG[i]))
			ok = false;
	}
	check(ok, label);
}

/* libcamera Rotation（逆时针）→ ANDROID_SENSOR_ORIENTATION（顺时针） */
static int32_t androidOrientation(int32_t rawRotation)
{
	int32_t ccw = ((rawRotation % 360) + 360) % 360;
	if (ccw % 90 != 0)
		ccw = 0;
	return (360 - ccw) % 360;
}

int main()
{
	printf("=== C. 运行级：编译并执行 deliverJpeg 的缩放/旋转路径 ===\n");
	for (int deg : { 0, 90, 180, 270 }) {
		testPixels(8, 6, 8, 6, deg, false);   /* 不缩放、只旋转（含快捷路径） */
		testPixels(8, 6, 4, 3, deg, true);    /* 等比缩小一半再旋转 */
		testPixels(8, 6, 4, 6, deg, true);    /* 非等比缩放 */
	}

	printf("\n=== D. 运行级：Rotation → ANDROID_SENSOR_ORIENTATION 换算 ===\n");
	struct { int32_t in, out; } cases[] = {
		/* in 是设备树/libcamera 的逆时针安装角，out 是 Android 的顺时针校正角 */
		{ 0, 0 }, { 90, 270 }, { 180, 180 }, { 270, 90 },
		{ 360, 0 },      /* 归一化 */
		{ -90, 90 },     /* -90 逆时针 ≡ 270 逆时针 ⇒ 校正角 90 */
		{ 45, 0 },       /* 非法 ⇒ 归 0 */
	};
	for (auto &c : cases) {
		char label[128];
		snprintf(label, sizeof(label), "Rotation=%d -> %d", c.in,
			 androidOrientation(c.in));
		check(androidOrientation(c.in) == c.out, label);
	}

	printf("\n=== 汇总 ===\n  检查项 %d，失败 %d\n", g_checks, g_fail);
	return g_fail ? 1 : 0;
}

File: scripts/camera/test-jpeg-rotation.py
#!/usr/bin/env python3
"""照片方向链路的离线自检（无依赖，纯 Python 3）。

为什么要写这个
--------------

整条链路里有【三个】互相独立的「90 度」，任何一个方向反了，用户看到的都是
「照片逆时针转了 90 度」这同一种症状 —— 而单看照片分辨不出是哪一个反了：

    ① 设备树 rotation                  逆时针，传感器物理安装角
    ② libcamera properties::Rotation   逆时针，继承 ①
    ③ ANDROID_SENSOR_ORIENTATION       顺时针，= 360 - ②
    ④ ANDROID_JPEG_ORIENTATION         顺时针，应用按当前设备姿态算出来的
    ⑤ libyuv RotationMode              顺时针，与 ④ 同向

HAL 对 ④ 的合同只有一条，也是本脚本 B 部分唯一断言的东西：
**把传感器读出顺时针旋转 ④ 度**。设备姿态怎么折算成 ④ 是应用的事，HAL 不该
再猜一次 —— 猜错就是双重校正，而且猜错了从照片上一样看不出来。

用法::

    python3 scripts/camera/test-jpeg-rotation.py
    python3 scripts/camera/test-jpeg-rotation.py --verbose

退出码 0 = 全部通过。
"""

import argparse
import os
import re
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
CAM_DIR = os.path.normpath(
    os.path.join(HERE, "..", "..", "device", "huawei", "gaokun3", "camera"))

FAILURES = []
CHECKS = 0


def check(cond, label, detail=""):
    global CHECKS
    CHECKS += 1
    if cond:
        print("  [PASS] %s" % label)
        if detail:
            print("         %s" % detail)
    else:
        print("  [FAIL] %s" % label)
        if detail:
            print("         %s" % detail)
        FAILURES.append(label)
    return cond


def read(name):
    with open(os.path.join(CAM_DIR, name), encoding="utf-8") as f:
        return f.read()


def strip_comments(text):
    """去掉注释 —— 免得断言被注释里的说明文字骗过去。"""
    text = re.sub(r"/\*.*?\*/", "", text, flags=re.S)
    text = re.sub(r"//[^\n]*", "", text)
    return text


# ─────────────────────────── A. 源码级断言 ───────────────────────────


def part_a():
    print("\n=== A. 源码级断言（解析 HAL 的 .cpp）===")

    session = read("Session.cpp")
    device = read("Device.cpp")
    header = read("Session.h")
    code = strip_comments(session)
    dev_code = strip_comments(device)

    # ── A1. ANDROID_JPEG_ORIENTATION 真的被读了 ──
    check("ANDROID_JPEG_ORIENTATION" in code,
          "A1 请求里的 ANDROID_JPEG_ORIENTATION 被消费",
          "老代码一个字都没读它，JPEG 永远停在传感器原始朝向")
    check(re.search(r"requestEntryInt\(\s*pend\.settings\s*,\s*ANDROID_JPEG_ORIENTATION",
                    code) is not None,
          "A1b 取值来源是【本帧的请求设置】，不是静态元数据")
    check("ANDROID_JPEG_QUALITY" in code,
          "A1c ANDROID_JPEG_QUALITY 也读进来了（顺手修掉质量恒 90）")

    # ── A2. libyuv 映射表必须是恒等映射（顺时针 → 顺时针） ──
    mapper = re.search(r"toLibyuvRotation\s*\(int32_t\s+\w+\)(.*?)\n\}", code, re.S)
    if not check(mapper is not None, "A2 找到 toLibyuvRotation() 的定义"):
        return
    body = mapper.group(1)
    table = {int(a): int(b)
             for a, b in re.findall(
                 r"case\s+(\d+)\s*:\s*return\s+libyuv::kRotate(\d+)\s*;", body)}
    check(table == {0: 0, 90: 90, 180: 180, 270: 270},
          "A2b 映射表是恒等的：Android 角 == libyuv 角",
          "实测表 = %s" % sorted(table.items()))
    check("360" not in body,
          "A2c 映射表里没有 360-x 之类的反向换算",
          "libyuv 的 kRotate90 本身就是顺时针，取反会转反 180 度")
    default_branch = re.search(r"default\s*:(.*?)(?=\n\s*case\s+\d+\s*:|\Z)", body, re.S)
    check(default_branch is not None and "kRotate0" in default_branch.group(1),
          "A2d 非法角度（非 0/90/180/270）落到 kRotate0",
          "Android 只允许这四个值，静默接受别的值会得到随机方向")

    # ── A3. 90/270 的宽高互换 ──
    compact = re.sub(r"\s+", "", code)
    check("swapDims?dstH:dstW" in compact and "constint32_toutW" in compact,
          "A3 90/270 时交换输出宽高：outW=dstH, outH=dstW",
          "BLOB 流只声明字节数，JPEG 实际尺寸换了是允许的")

    # ── A4. ARGBRotate 的目标 stride 必须用旋转后的宽度 ──
    rot_call = re.search(r"libyuv::ARGBRotate\((.*?)\)\s*!=\s*0", code, re.S)
    if check(rot_call is not None, "A4 找到 ARGBRotate 的调用"):
        args = [a.strip() for a in re.sub(r"\s+", "", rot_call.group(1)).split(",")]
        check(len(args) == 7 and args[3] == "outW*4",
              "A4b 目标 stride 用【旋转后】的宽度 outW*4",
              "实测第 4 个实参 = %r；沿用 dstW*4 会让每一行错位" % (args[3],))
        check(len(args) == 7 and args[4] == "dstW" and args[5] == "dstH",
              "A4c 源宽高仍是传感器坐标系的 (dstW, dstH)")
        check(len(args) == 7 and args[6] == "rotMode",
              "A4d 旋转模式来自请求解析出的 rotMode")
    else:
        check(False, "A4b/c/d 无法校验（没找到调用）")

    # ── A5. 实参顺序：方向必须真的传进编码函数 ──
    # 第一个实参是【源缓冲】：修前是 rgb，远端加入静态照片降噪后变成 rgbStill。
    # 名字不是合同（缓冲换谁进来都行）—— 合同是后两个实参必须是质量与方向。
    call = re.search(r"deliverJpeg\(\s*[A-Za-z_][A-Za-z0-9_]*\s*,(.*?)\)", code, re.S)
    if check(call is not None, "A5 找到 deliverJpeg 的调用点"):
        args = [a.strip() for a in re.sub(r"\s+", " ", call.group(1)).split(",")]
        check(len(args) == 6 and args[4] == "jpegQuality" and args[5] == "jpegOrientation",
              "A5b 调用点依次传入 (quality, jpegOrientation)",
              "实测实参 = %s" % args)
    sig = re.search(r"bool\s+Session::deliverJpeg\((.*?)\)\s*\{", code, re.S)
    if check(sig is not None, "A6 找到实现签名"):
        sig_args = [a.strip() for a in re.sub(r"\s+", " ", sig.group(1)).split(",")]
        check(len(sig_args) == 7 and sig_args[6].endswith("jpegOrientation"),
              "A6b 实现的形参里有 jpegOrientation",
              "实测 = %s" % sig_args)
    hdr_sig = re.search(r"bool\s+deliverJpeg\((.*?)\)\s*;", header, re.S)
    check(hdr_sig is not None and "jpegOrientation" in hdr_sig.group(1),
          "A6c 头文件里的声明同步了")

    # ── A7. Device.cpp 的逆时针 → 顺时针换算 ──
    check(re.search(r"facts_\.orientation\s*=\s*\(360\s*-\s*ccw\)\s*%\s*360\s*;",
                    dev_code) is not None,
          "A7 ANDROID_SENSOR_ORIENTATION = (360 - 逆时针安装角) % 360",
          "上游 libcamera 的 src/android/camera_device.cpp 就是这个式子")
    check(re.search(r"facts_\.orientation\s*=\s*rot\s*\?\s*\*\s*rot\s*:\s*0\s*;",
                    dev_code) is None,
          "A7b 老的直接抄数值（rot ? *rot : 0）已经删掉",
          "那个写法在 90/270 上差 180 度")
    check(re.search(r"ccw\s*%\s*90\s*!=\s*0", dev_code) is not None,
          "A7c 非 90 倍数的安装角被拦下并告警",
          "Android 只认 0/90/180/270")


# ─────────────────────── B. 行为模型（顺时针定义） ───────────────────────


def rotate_cw(img, deg):
    """按【顺时针】旋转 deg 度（deg ∈ {0,90,180,270}）。

    这是「顺时针」的数学定义，不依赖任何第三方库的命名：
        90 度：  out[i][j] = in[H-1-j][i]
        180 度： out[i][j] = in[H-1-i][W-1-j]
        270 度： out[i][j] = in[j][W-1-i]
    """
    deg %= 360
    if deg == 0:
        return [row[:] for row in img]
    h, w = len(img), len(img[0])
    if deg == 90:
        return [[img[h - 1 - j][i] for j in range(h)] for i in range(w)]
    if deg == 180:
        return [[img[h - 1 - i][w - 1 - j] for j in range(w)] for i in range(h)]
    if deg == 270:
        return [[img[j][w - 1 - i] for j in range(h)] for i in range(w)]
    raise ValueError(deg)


def libyuv_rotate(img, mode_deg):
    """按 libyuv 的【口头契约】旋转。

    include/libyuv/rotate.h 原文::

        kRotate90  = 90,   // Rotate 90 degrees clockwise.
        kRotate270 = 270,  // Rotate 270 degrees clockwise.
        kRotateClockwise = 90, kRotateCounterClockwise = 270,   // 历史别名

    所以就是 rotate_cw。上游哪天改了方向，这里和 A2 的断言会一起红掉。
    """
    return rotate_cw(img, mode_deg)


def part_b(verbose=False):
    print("\n=== B. 行为模型：场景 → 传感器读出 → HAL 旋转 → JPEG ===")
    print("""
    三个量的关系（很容易混，这里一次说清）：
      · dt_rotation  设备树/libcamera 的角，【逆时针】，= 把画面转正所需的
                     逆时针角；它【在数值上】恰好等于"照片看起来顺时针歪了多少"
      · apparent_cw  照片看起来顺时针歪了多少      = dt_rotation
      · sensor_orient ANDROID_SENSOR_ORIENTATION，【顺时针】校正角
                     = (360 - apparent_cw) % 360  = (360 - dt_rotation) % 360
    校正角与"歪了多少"是【互补】关系 —— 这正是最容易写反的地方。
""")

    # 一张非对称的 3x4 场景图，每格一个唯一标签。
    # 非对称是关键：对称图案会让「转反 180 度」也蒙混过关。
    upright = [[10 * (r + 1) + (c + 1) for c in range(4)] for r in range(3)]
    if verbose:
        print("    正立场景(3 行 x 4 列):")
        for row in upright:
            print("      %s" % row)

    print("\n  -- B1/B2 设备自然朝向下：安装角 → 元数据 → 照片 --")
    for dt_rotation in (0, 90, 180, 270):
        # 传感器读出：没做任何校正时，照片看起来顺时针歪了 dt_rotation 度。
        apparent_cw = dt_rotation
        raw = rotate_cw(upright, apparent_cw)

        # ① HAL 把逆时针的安装角换算成顺时针的校正角（这才是 Android 要的）
        sensor_orientation = (360 - dt_rotation) % 360
        check(sensor_orientation == (360 - apparent_cw) % 360,
              "B1 Rotation=%3d 度(逆时针) -> 上报 SENSOR_ORIENTATION=%3d 度(顺时针)"
              % (dt_rotation, sensor_orientation))

        # ② 设备自然朝向下，应用算出的 JPEG 旋转角就等于 SENSOR_ORIENTATION
        requested = sensor_orientation
        # ③ HAL 按 libyuv 顺时针语义旋转（A2 已断言映射恒等）
        out = libyuv_rotate(raw, requested)

        check(out == upright,
              "B2 Rotation=%3d 度: 修复后 JPEG == 正立场景" % dt_rotation)

        # 回归：老代码完全不动像素
        if dt_rotation == 0:
            check(raw == upright, "B3 Rotation=  0 度: 不旋转本来就对（老代码无辜）")
        else:
            check(raw != upright,
                  "B3 Rotation=%3d 度: 修复前 JPEG != 正立场景（用户看到的歪）"
                  % dt_rotation)
        if dt_rotation == 270:
            # 需要补 90 度顺时针而一点没补 ⇒ 看起来是 270 度顺时针
            # = 90 度逆时针，逐字对上用户的描述。
            check(raw == rotate_cw(upright, 270),
                  "B3b Rotation=270 度: 修复前看起来顺时针歪 270 度",
                  "270 度顺时针 == 90 度逆时针，这就是用户报的症状")

        # 宽高：校正角为 90/270 时必须换过来
        h, w = len(raw), len(raw[0])
        swapped = sensor_orientation in (90, 270)
        dst_w, dst_h = w, h                      # 流声明 = 传感器坐标系尺寸
        out_w = dst_h if swapped else dst_w      # HAL 的 outW
        out_h = dst_w if swapped else dst_h      # HAL 的 outH
        expect = (h, w) if swapped else (w, h)
        check((out_w, out_h) == expect,
              "B4 Rotation=%3d 度: 输出 %dx%d（流声明 %dx%d）"
              % (dt_rotation, out_w, out_h, dst_w, dst_h))

    # ── B5. HAL 对请求角的合同：忠实按【顺时针】转，不自己再猜设备姿态 ──
    print("\n  -- B5 HAL 的合同：JPEG == 传感器读出顺时针转 请求角 --")
    for dt_rotation in (0, 90, 180, 270):
        raw = rotate_cw(upright, dt_rotation)
        correct_angle = (360 - dt_rotation) % 360
        for requested in (0, 90, 180, 270):
            out = libyuv_rotate(raw, requested)
            faithful = out == rotate_cw(raw, requested)
            # 只有请求角正好是校正角时才该正立；任何别的组合都必须【不正立】，
            # 否则说明发生了双重校正。
            should_be_upright = (requested == correct_angle)
            check(faithful and (out == upright) == should_be_upright,
                  "B5 Rotation=%3d 请求=%3d: 忠实旋转，且仅角度正确时正立"
                  % (dt_rotation, requested),
                  "" if should_be_upright
                  else "此时本就不该正立（HAL 不替应用猜设备姿态）")

    # ── B6. 一次旋转 != 两次旋转（防止"转像素 + 又写 EXIF 角度"的经典错） ──
    print("\n  -- B6 不得双重校正 --")
    for dt_rotation in (90, 270):
        raw = rotate_cw(upright, dt_rotation)
        angle = (360 - dt_rotation) % 360
        once = libyuv_rotate(raw, angle)
        twice = rotate_cw(once, angle)
        check(once == upright and twice != upright,
              "B6 Rotation=%3d: 只转一次正好，转两次就歪" % dt_rotation,
              "所以 EXIF 里不能再写这个角，否则看图器会再转一遍")


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--verbose", action="store_true", help="打印场景矩阵")
    args = ap.parse_args()

    print("照片方向链路自检 —— 目录 %s" % CAM_DIR)
    part_a()
    part_b(args.verbose)

    print("\n=== 汇总 ===")
    print("  检查项 %d，失败 %d" % (CHECKS, len(FAILURES)))
    for f in FAILURES:
        print("  [X] %s" % f)
    if FAILURES:
        return 1
    print("  全部通过")
    return 0


if __name__ == "__main__":
    sys.exit(main())
